validator_client/src/duties_service/sync.rs

```diff
@@ -487,77 +489,75 @@ pub fn fill_in_aggregation_proofs<T: SlotClock + 'static, E: EthSpec>(
 
     // Generate selection proofs for each validator at each slot, one epoch at a time.
     for epoch in (current_epoch.as_u64()..=pre_compute_epoch.as_u64()).map(Epoch::new) {
-        // Generate proofs.
-        let validator_proofs: Vec<(u64, Vec<_>)> = pre_compute_duties
-            .iter()
-            .filter_map(|(validator_start_epoch, duty)| {
-                // Proofs are already known at this epoch for this validator.
-                if epoch < *validator_start_epoch {
-                    return None;
-                }
+        let mut validator_proofs = vec![];
+        for (validator_start_epoch, duty) in pre_compute_duties {
+            // Proofs are already known at this epoch for this validator.
+            if epoch < *validator_start_epoch {
+                continue;
+            }
 
-                let subnet_ids = duty
-                    .subnet_ids::<E>()
-                    .map_err(|e| {
-                        crit!(
+            let subnet_ids = match duty.subnet_ids::<E>() {
+                Ok(subnet_ids) => subnet_ids,
+                Err(e) => {
+                    crit!(
+                        log,
+                        "Arithmetic error computing subnet IDs";
+                        "error" => ?e,
+                    );
+                    continue;
+                }
+            };
+
+            let mut proofs = vec![];
+            for (duty_slot, &subnet_id) in epoch
+                .slot_iter(E::slots_per_epoch())
+                .cartesian_product(&subnet_ids)
+            {
+                // Construct proof for prior slot.
+                let slot = duty_slot - 1;
+
+                let proof = if let Ok(proof) = duties_service
+                    .validator_store
+                    .produce_sync_selection_proof(&duty.pubkey, slot, subnet_id)
+                    .await
```

Comment:
These futures should probably also be batched in parallel, as there are potentially quite a lot of them: `num_committee_members * 64 (two epochs lookahead) * 1-4 subnets`. On large networks this is less likely to be an issue, but could be troublesome on testnets with a large fraction of local validators